File: src/main/java/com/EmosewaPixel/pixellib/recipes/AbstractRecipeBuilder.java
package com.EmosewaPixel.pixellib.recipes;

import com.EmosewaPixel.pixellib.PixelLib;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public abstract class AbstractRecipeBuilder<T extends SimpleMachineRecipe, R extends AbstractRecipeBuilder<T, R>> {
    private AbstractRecipeList<T, R> recipeList;
    private List<Object> inputs = new ArrayList<>();
    private List<Object> outputs = new ArrayList<>();
    private int time = 0;

    public AbstractRecipeBuilder(AbstractRecipeList<T, R> recipeList) {
        this.recipeList = recipeList;
    }

    protected List<Object> getInputs() {
        return inputs;
    }

    protected List<Object> getOutputs() {
        return outputs;
    }

    protected int getTime() {
        return time;
    }

    protected AbstractRecipeList<T, R> getRecipeList() {
        return recipeList;
    }

    public R input(Object... inputs) {
        this.inputs.addAll(Arrays.asList(inputs));
        return (R) this;
    }

    public R output(Object... outputs) {
        this.outputs.addAll(Arrays.asList((outputs)));
        return (R) this;
    }

    public R time(int amount) {
        time = amount;
        return (R) this;
    }

    public abstract T build();

    public void buildAndRegister() {
        if (!build().isEmpty())
            recipeList.add(build());
        else
            PixelLib.LOGGER.error("Recipe with output {} is empty", outputs);
    }
}

File: src/main/java/com/EmosewaPixel/pixellib/recipes/AbstractRecipeBuilder.kt
package com.EmosewaPixel.pixellib.recipes

import com.EmosewaPixel.pixellib.PixelLib
import java.util.*

//Recipe Builders are builders used for easily creating Machine Recipes
abstract class AbstractRecipeBuilder<T : SimpleMachineRecipe, R : AbstractRecipeBuilder<T, R>>(protected val recipeList: AbstractRecipeList<T, R>) {
    protected val inputs = ArrayList<Any>()
    protected val consumeChances = ArrayList<Float>()
    protected val outputs = ArrayList<Any>()
    protected val outputChances = ArrayList<Float>()
    protected var time = 0
        private set

    fun input(vararg inputs: Any): R {
        this.inputs.addAll(listOf(*inputs))
        repeat(inputs.size) { consumeChances.add(1f) }
        return this as R
    }

    fun potentiallyConsumable(input: Any, consumeChance: Float): R {
        this.inputs.add(input)
        consumeChances.add(consumeChance)
        return this as R
    }

    fun notConsumable(vararg inputs: Any): R {
        this.inputs.addAll(listOf(*inputs))
        repeat(inputs.size) { consumeChances.add(0f) }
        return this as R
    }

    fun output(vararg outputs: Any): R {
        this.outputs.addAll(listOf(*outputs))
        repeat(outputs.size) { outputChances.add(1.0f) }
        return this as R
    }

    fun chancedOutput(output: Any, chance: Float): R {
        this.outputs.add(output)
        outputChances.add(chance)
        return this as R
    }

    fun time(amount: Int): R {
        time = amount
        return this as R
    }

    abstract fun build(): T

    fun buildAndRegister() {
        if (!build().isEmpty)
            recipeList.add(build())
        else
            PixelLib.LOGGER.error("Recipe with output {} is empty", outputs)
    }
}
